File: src/commands/sources/docker.ts
import Docker from "dockerode";
import select from "@inquirer/select";
import shell from "shelljs";
import fs from "fs";

import { Source } from ".";
import { StateManager } from "../../utils/stateManager";
import { writeToPath } from "../../utils/write";
import ora from "ora";

const generateConfig = (
  source: string,
  includeContainer: string,
  project: string,
  feed: string,
  token: string
) => `
${source.trim()}
include_containers = [ "${includeContainer}" ]

[sinks.lawg_sink]
type = "http"
encoding.codec = "json"
inputs = ["source0"]
uri = "http://100.127.114.55:8080/v1/projects/${project}/feeds/${feed}/logs"
auth.strategy = "bearer"
auth.token = "${token}"
`;

export const docker: Source = async (feed, project) => {
  const docker = new Docker();
  const fetchContainers = ora("Fetching containers").start();
  const containers = await docker.listContainers();
  fetchContainers.succeed("Fetched containers");

  const selectedContainer = await select({
    message: "Select a container",
    choices: containers.map((container) => ({
      name: `${container.Names[0].substring(1)} (${container.Id.substring(
        0,
        12
      )})`,
      value: container.Names[0].substring(1),
    })),
  });

  let vectorGeneratedConfig = shell.exec("vector generate docker_logs", {
    silent: true,
  }).stdout;

  // remove first 2 lines of generatedConfig (irrelevant to config) (im waiting for the day this is added https://tc39.es/proposal-pipeline-operator/)
  const lines = vectorGeneratedConfig.split("\n");
  lines.splice(0, 2)[0];

  vectorGeneratedConfig = lines.join("\n");

  const config = generateConfig(
    vectorGeneratedConfig,
    selectedContainer,
    project.namespace,
    feed.name,
    (await StateManager.getState()).token
  );

  const configPath = await writeToPath(
    `configs/${feed.name}-${selectedContainer}.toml`,
    config
  );
  console.log(`Config generated and saved to ${configPath}`);
};

File: commands/sources/docker.go
package sources

import (
	"context"
	"fmt"
	"leaf/utils"
	"os/exec"
	"strings"

	"github.com/AlecAivazis/survey/v2"
	"github.com/docker/docker/api/types"
	"github.com/docker/docker/client"
)

func DockerLogs(feed utils.Feed, project utils.Project) error {
	client, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
	if err != nil {
		return utils.ParsedError(err, "Failed to connect to Docker daemon", true)
	}

	containers, err := client.ContainerList(context.Background(), types.ContainerListOptions{})

	if err != nil {
		return utils.ParsedError(err, "Failed to list containers", true)
	}

	var containerNames = []string{}
	for _, container := range containers {
		containerNames = append(containerNames, fmt.Sprintf("%s (%s)", container.Names[0][1:], container.ID[:12]))
	}

	var selectedContainer string
	survey.AskOne(&survey.Select{
		Message: "Select a container:",
		Options: containerNames,
	}, &selectedContainer, survey.WithValidator(survey.Required))

	selectedContainer = strings.Split(selectedContainer, " ")[0]

	cmd := exec.Command("vector", "generate", "docker_logs")
	vectorGeneratedConfig, err := cmd.Output()
	if err != nil {
		return utils.ParsedError(err, "Failed to generate config", true)
	}

	stringedVectorGeneratedConfig := string(vectorGeneratedConfig[:])

	lines := strings.Split(stringedVectorGeneratedConfig, "\n")
	lines = lines[2:] // remove first 2 lines of generatedConfig (irrelevant to config)

	stringedVectorGeneratedConfig = strings.Join(lines, "\n")

	if err := utils.GenerateConfig(fmt.Sprintf(`%s
include_containers = [ "%s" ]`, stringedVectorGeneratedConfig, selectedContainer), fmt.Sprintf("%s-%s-docker-logs", feed.Name, selectedContainer), project.Namespace, feed.Name); err != nil {
		return utils.ParsedError(err, "Failed to generate config", true)
	}

	return nil
}
